subxt/src/events.rs

```diff
@@ -38,17 +37,357 @@ use codec::{
+        &self,
+    ) -> impl Iterator<Item = Result<EventDetails<Evs>, BasicError>> + '_ {
+        let event_bytes = &self.event_bytes;
+
+        let mut pos = 0;
+        let mut index = 0;
+        std::iter::from_fn(move || {
+            let cursor = &mut &event_bytes[pos..];
+            let start_len = cursor.len();
+
+            if start_len == 0 || self.num_events == index {
+                None
+            } else {
+                let mut decode_one_event = || -> Result<_, BasicError> {
+                    let phase = Phase::decode(cursor)?;
+                    let ev = Evs::decode(cursor)?;
+                    let _topics = Vec::<T::Hash>::decode(cursor)?;
+                    Ok((phase, ev))
+                };
+                match decode_one_event() {
+                    Ok((phase, event)) => {
+                        // Skip over decoded bytes in next iteration:
+                        pos += start_len - cursor.len();
+                        // Gather the event details before incrementing the index for the next iter.
+                        let res = Some(Ok(EventDetails {
+                            phase,
+                            index,
+                            event,
+                        }));
+                        index += 1;
+                        res
+                    }
+                    Err(e) => {
+                        // Next iteration will return None:
+                        pos = event_bytes.len();
+                        Some(Err(e))
+                    }
+                }
+            }
+        })
+    }
+
+    /// Iterate over all of the events, using metadata to dynamically
+    /// decode them as we go, and returning the raw bytes and other associated
+    /// details.
+    ///
+    /// This method is safe to use even if you do not statically know about
+    /// all of the possible events; it splits events up using the metadata
+    /// obtained at runtime, which does.
+    pub fn iter_raw(
+        &self,
+    ) -> impl Iterator<Item = Result<RawEventDetails, BasicError>> + '_ {
+        let event_bytes = &self.event_bytes;
+
+        let mut pos = 0;
+        let mut index = 0;
+        std::iter::from_fn(move || {
+            let cursor = &mut &event_bytes[pos..];
+            let start_len = cursor.len();
+
+            if start_len == 0 || self.num_events == index {
+                None
+            } else {
+                match decode_raw_event_details::<T>(self.metadata, index, cursor) {
+                    Ok(raw_event) => {
+                        // Skip over decoded bytes in next iteration:
+                        pos += start_len - cursor.len();
+                        // Increment the index:
+                        index += 1;
+                        // Return the event details:
+                        Some(Ok(raw_event))
+                    }
+                    Err(e) => {
+                        // Next iteration will return None:
```

Comment:
Okay it took me a minute to see why this is the case, we are triggering the `start_len == 0` condition. Maybe improve this comment or use an explicit mut bool flag (unless we do need to move the cursor the end anyway)

Also would be good to see a test for this as per my other comment.